### PR TITLE
✨ data-amp-bind-dynamic-class for AMP Bind

### PR DESCRIPTION
**Pull request by kristoferbaxter**
Allow authors to specify static and dynamic parts of the class syntax per element with AMP Bind.

Remaining work:
- [ ] Make `[class]` and `data-amp-bind-dynamic-class` attributes mutually exclusive via validation rules. Authors must choose one per element.
- [ ] Additional tests, with toggling of classes via other means.

### CODE REVIEW

File: extensions/amp-bind/0.1/test/integration/test-bind-impl.js [hunk 1]
```diff
@@ -750,6 +762,18 @@ describe
...
+        it('should support binding to CSS classes with arrays, including class+', () => {
+          const element = createElement(
+            env,
+            container,
+            "class='a' [class+]=\"['b','c']\""
+          );
+          expect(toArray(element.classList)).to.deep.equal(['a']);
+          return onBindReadyAndSetState(env, bind, {}).then(() => {
+            expect(toArray(element.classList)).to.deep.equal(['a', 'b', 'c']);
+          });
```
Comment (jridgewell) [inline extensions/amp-bind/0.1/test/integration/test-bind-impl.js:774]: Can you add a demo where `class+` is toggle-able?

Comment (kristoferbaxter) [reply extensions/amp-bind/0.1/test/integration/test-bind-impl.js:774]: Yes, a lot more tests are needed. I'll note this in the PR description.

File: extensions/amp-bind/0.1/bind-impl.js [hunk 2]
```diff
@@ -415,9 +415,9 @@ export class Bind {
     }
     // Only pass state for history updates to trusted viewers, since they
     // may contain user data e.g. form input.
-    return this.viewer_.isTrustedViewer().then((trusted) => {
```
Comment (kristoferbaxter) [inline extensions/amp-bind/0.1/bind-impl.js:418]: Self Review: This change isn't relevant. Back it out.

File: extensions/amp-bind/0.1/bind-validator.js [hunk 3]
```diff
@@ -36,6 +36,9 @@ const GLOBAL_PROPERTY_RULES = {
   'class': {
     denylistedValueRegex: '(^|\\W)i-amphtml-',
   },
+  'class+': {
+    denylistedValueRegex: '(^|\\W)i-amphtml-',
```
Comment (kristoferbaxter) [inline extensions/amp-bind/0.1/bind-validator.js:40]: Self review, these literals are the same. Consider if there is a better pattern.

File: validator/validator-main.protoascii [hunk 4]
```diff
@@ -5819,6 +5819,7 @@ attr_lists: {
   attrs: { name: "[aria-valuenow]" }
   attrs: { name: "[aria-valuetext]" }
   attrs: { name: "[class]" }
+  attrs: { name: "[class+]" }
```
Comment (kristoferbaxter) [inline validator/validator-main.protoascii:5822]: @honeybadgerdontcare The pattern I was trying here locally was...

```
attrs: {
  name: "[class]"
  trigger: {
    mandatory_oneof: "['[class]', '[class+]']"
  }
}
attrs: {
  name: "[class+]"
  trigger: {
    mandatory_oneof: "['[class]', '[class+]']"
  }
}
```

Going to likely need a custom specific rule here or the ability to indicate mutually exclusive attributes like this for global elements. Another option is to error in `amp-bind` for this scenario.

**Review by kristoferbaxter — COMMENTED** [review]
Leaving a few notes while this is being worked on.

File: validator/validator-main.protoascii (hunk 4 above)
Comment (honeybadgerdontcare) [reply validator/validator-main.protoascii:5822]: I'm inclined to something like `disallowed_attr` as a new field for attributes that would act similar to the trigger. The trigger was introduced depending on the value of the attribute. I don't think we need to expand that and could put this on the attr spec. Reviewing with wg-caching.

**Review by jridgewell — COMMENTED** [review]
Another issue this is going to is is `+` being invalid as part of an attribute name when getting/setting. We had a several issues related to the current `[X]` syntax for the same reason, and it led us recommend `data-amp-bind-X` pattern instead. I'm a little hesitant to add another non-standard name char for bind.

**Comment by kristoferbaxter** [comment]
@jridgewell Very good point, this doesn't need to use the current naming from the feature request.

I'll move this over to `data-amp-dynamic-classes` (unless someone else has a better name).

**Pull request closed without merging**